src/ShopBundle/Basket/BasketVariableReplacer.php

```diff
@@ -0,0 +1,140 @@
+<?php
+
+/*
+ * This file is part of the Chameleon System (https://www.chameleonsystem.com).
+ *
+ * (c) ESONO AG (https://www.esono.de)
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace ChameleonSystem\ShopBundle\Basket;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Twig\Environment;
+use Twig\Error\Error;
+
+/**
+ * BasketVariableReplacer is used to add all request parameters as hidden fields to "Add to Basket"-forms.
+ * This happens as a post process job to avoid caching those parameters.
+ *
+ * To successfully use it on a page the template for the basket button must include a placeholder in the form of:
+ *
+ * [{BASKETHIDDENFIELDS}]
+ *
+ * Background: the additional fields are added so that those parameters from the url do not get lost on the redirect
+ * after the POST call to the server.
+ * As we do not know which parameters might be needed by other modules on the page we must add all of them.
+ * This isn't a security concern as the page has already been called with those parameters in the first place.
+ */
+final class BasketVariableReplacer
+{
+    public const BASKET_HIDDEN_FIELDS_PLACEHOLDER = 'BASKETHIDDENFIELDS';
+    private const HIDDEN_FIELDS_SNIPPET = '@ChameleonSystemShop/snippets/ShopBundle/BasketForm/hiddenBasketFields.html.twig';
+
+    /**
+     * @var Environment
+     */
+    private $twigEnvironment;
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    public function __construct(
+        Environment $twigEnvironment,
+        LoggerInterface $logger)
+    {
+        $this->twigEnvironment = $twigEnvironment;
+        $this->logger = $logger;
+    }
+
+    /**
+     * handleRequest will be invoked on kernel.request and add the hidden fields to the replacer.
+     * On error it will log to the request channel and move on. It will not halt execution.
+     *
+     * @param GetResponseEvent $event
+     */
+    public function handleRequest(GetResponseEvent $event): void
+    {
+        $request = $event->getRequest();
+        $queryParameters = $request->query->all();
+        $paramsToRender = $this->filterKeys($queryParameters, ['basket']);
```

Comment:
I would like to replace the literal string 'basket' with the constant used for the parameter (\MTShopBasketCoreEndpoint::URL_REQUEST_PARAMETER). That couples this class to the basket class - but then again, so das the literal string.

Reply:
Done both. The coupling is ok in this case as the class itself only deals with the basket workflow in the first place.